File: codes/naeun/496_nextgreaterelement1_1.java
import java.util.*;

class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int ans[] = new int[nums1.length];
        Stack<Integer> stack = new Stack<>();

        int greter[] = new int[10001]; // Array to store next greater elements
        Arrays.fill(greter, -1); // Default to -1 (no greater element)

        // Traverse nums2 to find the next greater elements
        for (int i = 0; i < nums2.length; i++) {
            while (!stack.isEmpty() && nums2[stack.peek()] < nums2[i]) {
                int index = stack.pop();
                greter[nums2[index]] = nums2[i];
            }
            stack.push(i);
        }

        // Map the results from nums2 to nums1
        for (int i = 0; i < nums1.length; i++) {
            ans[i] = greter[nums1[i]];
        }
        return ans;
    }
}

File: codes/yujo/496_nextgreaterelement1_1.ts
function nextGreaterElement(nums1: number[], nums2: number[]): number[] {
  // Use Map to store the next greater element for each number
  const nextGreater = new Map<number, number>();
  // Monotonic decreasing stack
  const stack: number[] = [];

  // Iterate through nums2 to find next greater elements
  for (const num of nums2) {
    // If current number is greater than the top of stack,
    // it is the next greater element for the popped number
    while (stack.length > 0 && stack[stack.length - 1] < num) {
      nextGreater.set(stack.pop()!, num);
    }
    stack.push(num);
  }

  // Numbers remaining in stack have no next greater element
  while (stack.length > 0) {
    nextGreater.set(stack.pop()!, -1);
  }

  // Add next greater element for each number in nums1 to result array
  return nums1.map((num) => nextGreater.get(num)!);
}